client/consensus/grandpa/justification.go

```diff
@@ -0,0 +1,315 @@
+	}
+
+	return justification, justification.verifyWithVoterSet(setID, voters)
+}
+
+// TODO get feedback on if I can avoid this, see below
+//type NewAuthority[ID constraints.Ordered] struct {
+//	Key    ID
+//	Weight uint64
+//}
+
+// Validate the commit and the votes' ancestry proofs.
+func (j *Justification[H, N, S, ID, Header]) verify(setID uint64, weights []finalityGrandpa.IDWeight[ID]) error {
+	// TODO Get reviewer feedback on this. In substrate they pass in data then convert to IDWeight, however for
+	// us we do no ever use this input type, so I think better to just directly take IDWeights as a param.
+	// Can revert if people disagree
+	//var weights []finalityGrandpa.IDWeight[ID]
+	//for _, authority := range authorities {
+	//	weight := finalityGrandpa.IDWeight[ID]{
+	//		ID: authority.Key,
+	//	}
+	//	weights = append(weights, weight)
+	//}
+	voters := finalityGrandpa.NewVoterSet[ID](weights)
+	if voters != nil {
+		err := j.verifyWithVoterSet(setID, *voters)
+		return err
+	}
+	return fmt.Errorf("%w", errInvalidAuthoritiesSet)
+}
+
+// Validate the commit and the votes' ancestry proofs.
+func (j *Justification[H, N, S, ID, Header]) verifyWithVoterSet(
+	_ uint64,
+	voters finalityGrandpa.VoterSet[ID]) error {
+	ancestryChain := newAncestryChain[H, N](j.VotesAncestries)
+	commitValidationResult, err := finalityGrandpa.ValidateCommit[H, N, S, ID](j.Commit, voters, ancestryChain)
+	if err != nil {
+		return fmt.Errorf("%w: invalid commit in grandpa justification", errBadJustification)
+	}
+
+	if !commitValidationResult.Valid() {
+		return fmt.Errorf("%w: invalid commit in grandpa justification", errBadJustification)
+	}
+
+	// we pick the precommit for the lowest block as the base that
+	// should serve as the root block for populating ancestry (i.e.
+	// collect all headers from all precommit blocks to the base)
+	precommits := j.Commit.Precommits
+	var minPrecommit *finalityGrandpa.SignedPrecommit[H, N, S, ID]
+	if len(precommits) == 0 {
+		panic("can only fail if precommits is empty; commit has been validated above; " +
+			"valid commits must include precommits")
+	}
+	for _, precommit := range precommits {
+		currPrecommit := precommit
+		if minPrecommit == nil {
+			minPrecommit = &currPrecommit
+		} else if currPrecommit.Precommit.TargetNumber <= minPrecommit.Precommit.TargetNumber {
+			minPrecommit = &currPrecommit
+		}
+	}
+
+	baseHash := minPrecommit.Precommit.TargetHash
+	visitedHashes := make(map[H]struct{})
+	for _, signed := range precommits {
+		/*
+				TODO signature is generic type any, but signature verification in Gossamer uses []byte
+			    TODO need access to message.value both for encoding and so I can set it
+		*/
+		//if !checkMessageSignature[H, N](signed.Precommit, signed.ID, signed.Signature, j.Round, setID) {
+		//	return fmt.Errorf("%w: invalid signature for precommit in grandpa justification", errBadJustification)
+		//}
```

Comment:
doesn't this need to be done?

Reply:
Yes, however, I am currently struggling because the justification ID is constrained to be constraints.Ordered and for this we need to be able to cast to an ed25519.PubKey type, which is an invalid conversion.

I am still brainstorming how to approach this and would be happy to sync to discuss this issue

Reply:
This has been added and resolved now